### PR TITLE
(feat) Add a global SWR configuration

### PR DESCRIPTION
**Pull request by denniskigen**
# Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. Ensure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing#contributing-guidelines) label (such as `feat`, `fix`, or `chore`, among others). See existing PR titles for inspiration.

## For changes to apps

- [ ] My work conforms to the [**O3 Styleguide**](https://om.rs/styleguide) and [**design documentation**](https://om.rs/o3ui).

## If applicable

- [ ] My work includes tests or is validated by existing tests.
- [ ] I have updated the [esm-framework mock](https://github.com/openmrs/openmrs-esm-core/blob/main/packages/framework/esm-framework/mock.tsx) to reflect any API changes I have made.

## Summary

This PR adds a minimal [global SWR Configuration](https://swr.vercel.app/docs/global-configuration) to the `openmrsComponentDecorator` with some sensible defaults. These are:

```
const defaultSwrConfig = {
  errorRetryCount: 3,
  fetcher: openmrsFetch,
  focusThrottleInterval: 30000,
};
```

- `errorRetryCount` is the maximum number of retries to attempt after requests fail. SWR does not provide a default value out of the box AFAICT. 
- `focusThrottleInterval` - controls the throttling behaviour when revalidating stale data on focus. It sets a minimum interval between revalidation events when the window or tab regains focus. I've set it to 30000ms, which means that SWR will revalidate max once per 30 seconds of focus. The goal is to prevent load spikes from revalidation events but to keep the value low enough to get important updates. Defaults to 5 seconds.
- `fetcher` is the fetcher function to use.

An important things to note is that `SWRConfig` [merges](https://swr.vercel.app/docs/global-configuration#nesting-configurations) the configuration from the parent context. This means that alternative configurations provided further down the line will override the parent values.

For example, I've set the `refreshInterval` in the Active Visits frontend module to `500ms` here:

https://github.com/openmrs/openmrs-esm-core/assets/8509731/0f36566e-23a2-4b31-9bcb-0563159d6a6d

Consequently, revalidations are happening twice every second. All the other options from the global config are unaffected.

One other thing I've done here is cleaning up some of the type annotations for our custom error boundary component.

## TL;Dr

- SWR will retry failed requests for a maximum of three times.
- SWR will only revalidate on focus once every thirty seconds.
- The default fetcher is `openmrsFetch`.
- This top-level config can be overridden as `SWRConfig` merges configurations.
- These configuration options will apply for all SWR hooks unless configuration overrides are provided further downstream.

### CODE REVIEW

File: packages/framework/esm-react-utils/src/openmrsComponentDecorator.tsx [hunk 1]
```diff
@@ -1,15 +1,26 @@
...
   throwErrorsToConsole: true,
   disableTranslations: false,
 };
 
+// Read more about the available config options here: https://swr.vercel.app/docs/api#configuration
+const defaultSwrConfig = {
+  // max number of retries after requests have failed
+  errorRetryCount: 3,
+  // default fetcher function
+  fetcher: openmrsFetch,
```
Comment (vasharma05) [inline packages/framework/esm-react-utils/src/openmrsComponentDecorator.tsx:19]: If we don't pass the fetcher, which is `openmrsFetch` it will default to the `openmrsFetch`, right?

File: packages/framework/esm-react-utils/src/openmrsComponentDecorator.tsx [hunk 2]
```diff
@@ -1,15 +1,26 @@
...
+  // only revalidate once every 30 seconds
+  focusThrottleInterval: 30000,
```
Comment (vasharma05) [inline packages/framework/esm-react-utils/src/openmrsComponentDecorator.tsx:21]: This was really needed 🚀

File: packages/framework/esm-react-utils/src/openmrsComponentDecorator.tsx (hunk 1 above)
Comment (denniskigen) [reply packages/framework/esm-react-utils/src/openmrsComponentDecorator.tsx:19]: That's right.

**Review by ibacher — APPROVED** [review]

**Comment by denniskigen** [comment]
Good to go, @ibacher? Or should we wait for feedback from others?

**Comment by ibacher** [comment]
Best way to figure it out is to merge and see, I think.

**Comment by denniskigen** [comment]
Looks good. I see the network tab is very quiet.
